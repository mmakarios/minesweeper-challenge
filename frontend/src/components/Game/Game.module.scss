.game {
}

.afterGameActions {
  margin-top: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}
