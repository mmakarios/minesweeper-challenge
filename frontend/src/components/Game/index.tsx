export { default, BoardContext } from "./Game";
