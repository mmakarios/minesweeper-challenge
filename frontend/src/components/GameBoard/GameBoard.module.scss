.gameboard {
  display: flex;
  flex-direction: column;
}

.gameboardRow {
  display: flex;
}
