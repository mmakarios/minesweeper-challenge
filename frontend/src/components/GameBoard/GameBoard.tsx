import { useCallback, useContext, useMemo } from "react";
import { openBox } from "../../apis";
import BoxButton from "../BoxButton";
import { BoardContext } from "../Game";
import styles from "./GameBoard.module.scss";

export const GameBoard = () => {
  const { board, onBoardChange } = useContext(BoardContext);

  const onOpenBox = useCallback(
    async (box) => {
      console.log(`box`, box);
      // setIsLoading(true);
      // setAlert("");

      if (!board || !onBoardChange) {
        return;
      }
      let newBoard;
      try {
        newBoard = await openBox(board.id, box);
      } catch (err) {
        // setAlert("Board creation failed.");
      }
      // console.log(`newBoard`, newBoard);
      onBoardChange(newBoard);
      // setIsLoading(false);
      // history.push(`/${newBoard.id}`);
      // console.log(newBoard);
    },
    [board, onBoardChange]
  );

  const renderBoard = useMemo(() => {
    return board?.boxes.map((row, rowIndex) => (
      <div key={"row-" + rowIndex} className={styles.gameboardRow}>
        {row.map((box, columnIndex) => (
          <BoxButton
            onClick={() =>
              onOpenBox(columnIndex + rowIndex * board.boxes[0].length)
            }
            key={"box-" + columnIndex}
            box={box}
          />
        ))}
      </div>
    ));
  }, [board, onOpenBox]);

  return <div className={styles.gameboard}>{board && renderBoard}</div>;
};

export default GameBoard;
