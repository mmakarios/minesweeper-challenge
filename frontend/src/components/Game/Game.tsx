import { useCallback } from "react";
import styles from "./Game.module.scss";
import { BoardContextType, Board } from "../../types";
import { useEffect, useState, createContext } from "react";
import { retrieveBoard } from "../../apis";
import { useParams } from "react-router-dom";
import GameBoard from "../GameBoard";

export const BoardContext = createContext<BoardContextType>({});

export const Game = () => {
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [alert, setAlert] = useState<string>("");
  const [board, setBoard] = useState<Board>();
  const { id } = useParams<{ id: string }>();

  useEffect(() => {
    async function fetchBoard() {
      console.log(`fetching board`);
      setIsLoading(true);
      setAlert("");
      let board;
      try {
        board = await retrieveBoard(id);
      } catch (err) {
        setAlert("Error fetching the board!");
      }
      console.log("fetch board: ", board);
      setBoard(board);
      setIsLoading(false);
    }

    fetchBoard();
  }, [id]);

  const onBoardChange = useCallback(
    (board: Board) => {
      setBoard(board);
    },
    [setBoard]
  );

  return (
    <div className={styles.Game}>
      {isLoading && <div>Loading game...</div>}
      <BoardContext.Provider value={{ board, onBoardChange }}>
        <GameBoard />
      </BoardContext.Provider>
    </div>
  );
};

export default Game;
