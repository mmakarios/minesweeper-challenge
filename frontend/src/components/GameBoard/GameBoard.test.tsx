import React from "react";
import { act, create } from "react-test-renderer";
import GameBoard from "./GameBoard";
import { BoardContext } from "../Game";

it("renders component", () => {
  const tree = create(
    <BoardContext.Provider
      value={{
        board: {
          id: "id",
          status: 0,
          boxes: [
            [
              { value: "", state: "opened" },
              { value: "", state: "unopened" },
            ],
            [
              { value: "1", state: "opened" },
              { value: "m", state: "unopened" },
            ],
          ],
          started_at: "2021-03-10 07:38:12",
        },
      }}
    >
      <GameBoard />
    </BoardContext.Provider>
  );
  const buttons = tree.root.findAllByType("button");
  act(() => {
    buttons.forEach((button) => {
      button.props.onClick();
    });
  });
  expect(tree.toJSON()).toMatchSnapshot();
});
