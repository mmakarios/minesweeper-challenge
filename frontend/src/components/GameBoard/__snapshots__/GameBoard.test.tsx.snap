// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
<div
  className="gameboard"
>
  <div
    className="gameboardRow"
  >
    <button
      className="boxbutton opened undefined"
      disabled={true}
      onClick={[Function]}
      onContextMenu={[Function]}
    >
      
    </button>
    <button
      className="boxbutton unopened undefined"
      disabled={false}
      onClick={[Function]}
      onContextMenu={[Function]}
    />
  </div>
  <div
    className="gameboardRow"
  >
    <button
      className="boxbutton opened one"
      disabled={true}
      onClick={[Function]}
      onContextMenu={[Function]}
    >
      1
    </button>
    <button
      className="boxbutton unopened undefined"
      disabled={false}
      onClick={[Function]}
      onContextMenu={[Function]}
    >
      ðŸ’£
    </button>
  </div>
</div>
`;
